File: 024-spark-test/src/main/java/com/demien/sparktest/util/RestTestUtil.java
package com.demien.sparktest.util;

import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;

public class RestTestUtil {

    public static class RequestResult {

        public final String body;
        public final int status;

        private RequestResult(int status, String body) {
            this.body = body;
            this.status = status;
        }
    }

    public static RequestResult sendRequest(String method, String path, String urlParameters) throws IOException {

        URL url = new URL(path);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setDoOutput(true);
        conn.setInstanceFollowRedirects(false);
        conn.setRequestMethod(method);

        if (urlParameters!=null) {
            byte[] postData = urlParameters.getBytes(StandardCharsets.UTF_8);
            int postDataLength = postData.length;

            conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
            conn.setRequestProperty("charset", "utf-8");
            conn.setRequestProperty("Content-Length", Integer.toString(postDataLength));
            conn.setUseCaches(false);
            conn.getOutputStream().write(postData);
        }

        Reader in = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
        StringBuilder sb = new StringBuilder();
        for (int c; (c = in.read()) >= 0; )
            sb.append((char) c);
        String responseBody = sb.toString();
        int responseCode=conn.getResponseCode();
        return new RequestResult(responseCode, responseBody);

    }

    public static RequestResult sendRequest(String method, String path) throws IOException {
        return sendRequest(method, path, null);
    }


}

File: 047-money-transfer/src/test/scala/com/demien/mtransfer/rest/RestTestUtil.scala
package com.demien.mtransfer.rest

import java.io.{BufferedReader, InputStreamReader}
import java.net.{HttpURLConnection, URL}
import java.nio.charset.StandardCharsets

object RestTestUtil {

  class RequestResult(val status: Int, val body: String)


  def sendRequest(method: String, path: String, urlParameters: String): RequestResult = {
    val url = new URL(path)
    val conn = url.openConnection.asInstanceOf[HttpURLConnection]
    conn.setDoOutput(true)
    conn.setInstanceFollowRedirects(false)
    conn.setRequestMethod(method)
    if (urlParameters != null) {
      val postData = urlParameters.getBytes(StandardCharsets.UTF_8)
      val postDataLength = postData.length
      conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded")
      conn.setRequestProperty("charset", "utf-8")
      conn.setRequestProperty("Content-Length", Integer.toString(postDataLength))
      conn.setUseCaches(false)
      conn.getOutputStream.write(postData)
    }
    val in = new BufferedReader(new InputStreamReader(conn.getInputStream, "UTF-8"))
    val sb = new StringBuilder
    var c = 0

    do {
      c = in.read
      sb.append(c.toChar)
    } while (c > 0)

    in.close()


    val responseBody = sb.toString.replaceAll("[^\\x00-\\x7F]", "")
    val responseCode = conn.getResponseCode
    new RequestResult(responseCode, responseBody)
  }

  def sendRequest(method: String, path: String): RequestResult = sendRequest(method, path, null)

}
